File: smart-token-server.py
import os
import time
import json
import sys
import threading
from datetime import datetime
from flask import Flask, jsonify, request
import logging
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.common.exceptions import TimeoutException, WebDriverException
import requests

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = Flask(__name__)

class SeleniumTokenServer:
    def __init__(self):
        self.key_id = os.getenv('KEY_ID', 'KEY-8GFN9U3L0U')
        self.current_token = None
        self.token_info = None
        self.last_update = 0
        self.is_fetching = False
        self.auto_refresh = False
        self.refresh_interval = 900  # 15 minutes default
        self.cooldown_until = 0
        self.driver = None

    def log_with_time(self, message):
        now = datetime.now()
        time_str = now.strftime('%H:%M:%S')
        date_str = now.strftime('%d/%m/%Y')
        log_message = f'[{time_str} {date_str}] {message}'
        logger.info(log_message)
        print(log_message)

    def send_telegram_message(self, token):
        try:
            bot_token = os.getenv('TELEGRAM_BOT_TOKEN')
            chat_id = os.getenv('TELEGRAM_CHAT_ID')

            if not bot_token or not chat_id:
                self.log_with_time("⚠️ Telegram credentials not configured")
                return False

            message = f"🎉 **Token mới được lấy thành công!**\n\n"
            message += f"🔑 **Token:** `{token}`\n"
            message += f"⏰ **Thời gian:** {datetime.now().strftime('%H:%M:%S %d/%m/%Y')}\n"
            message += f"🌐 **Server:** Selenium Token Server"

            url = f"https://api.telegram.org/bot{bot_token}/sendMessage"
            data = {
                'chat_id': chat_id,
                'text': message,
                'parse_mode': 'Markdown'
            }

            response = requests.post(url, data=data, timeout=10)
            if response.status_code == 200:
                self.log_with_time("✅ Token sent to Telegram successfully")
                return True
            else:
                self.log_with_time(f"❌ Failed to send to Telegram: {response.status_code}")
                return False

        except Exception as e:
            self.log_with_time(f"❌ Telegram error: {str(e)}")
            return False

    def setup_browser(self):
        """Setup Chrome browser"""
        self.log_with_time('🌐 Setting up browser...')
        
        chrome_options = Options()
        chrome_options.add_argument("--no-sandbox")
        chrome_options.add_argument("--disable-dev-shm-usage")
        chrome_options.add_argument("--disable-gpu")
        chrome_options.add_argument("--headless=new")
        chrome_options.add_argument("--disable-web-security")
        chrome_options.add_argument("--window-size=1920,1080")
        chrome_options.add_argument("--disable-extensions")
        chrome_options.add_argument("--disable-plugins")
        chrome_options.add_argument("--disable-images")
        chrome_options.add_argument("--no-first-run")
        chrome_options.add_argument("--disable-background-timer-throttling")
        chrome_options.add_argument("--disable-backgrounding-occluded-windows")
        chrome_options.add_argument("--disable-renderer-backgrounding")
        
        try:
            self.driver = webdriver.Chrome(options=chrome_options)
            self.driver.set_page_load_timeout(30)
            self.driver.implicitly_wait(10)
            self.log_with_time('✅ Browser ready')
            return True
        except Exception as e:
            self.log_with_time(f'❌ Browser setup failed: {e}')
            return False

    def login_and_get_token(self):
        """Login and get token using Selenium"""
        if not self.driver:
            if not self.setup_browser():
                return {'success': False, 'error': 'Browser setup failed'}

        try:
            self.log_with_time('🔑 Logging into tokencursor.io.vn...')
            
            # Navigate to login page
            self.driver.get('https://tokencursor.io.vn/login')
            self.log_with_time('📡 Login page loaded')
            time.sleep(3)
            
            # Find key input field
            key_input = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.CSS_SELECTOR, 'input[name="key"]'))
            )
            self.log_with_time('📝 Found key input field')
            
            # Clear and fill key
            key_input.clear()
            key_input.send_keys(self.key_id)
            self.log_with_time(f'⌨️ Entered key: {self.key_id}')
            
            # Find and click login button
            login_button = WebDriverWait(self.driver, 10).until(
                EC.element_to_be_clickable((By.CSS_SELECTOR, 'button[type="submit"]'))
            )
            self.log_with_time('🔘 Found login button')
            
            # Click login button
            login_button.click()
            self.log_with_time('🔄 Clicking login button...')
            time.sleep(5)
            
            # Check if redirected to app page
            current_url = self.driver.current_url
            if 'app' in current_url:
                self.log_with_time('✅ Login successful, on app page')
                
                # Handle any popups
                try:
                    understood_button = WebDriverWait(self.driver, 5).until(
                        EC.element_to_be_clickable((By.XPATH, '//button[contains(text(), "Đã hiểu")]'))
                    )
                    understood_button.click()
                    self.log_with_time('✅ Clicked "Đã hiểu" button')
                    time.sleep(2)
                except TimeoutException:
                    self.log_with_time('⚠️ No popup found')
                
                # Look for token button with multiple possible texts
                try:
                    # Try different button texts
                    button_texts = [
                        '//button[contains(text(), "Lấy Token")]',
                        '//button[contains(text(), "Get Token")]',
                        '//button[contains(text(), "Token")]',
                        '//button[contains(text(), "Lấy")]',
                        '//button[contains(text(), "Get")]'
                    ]
                    
                    get_token_button = None
                    for button_text in button_texts:
                        try:
                            get_token_button = WebDriverWait(self.driver, 5).until(
                                EC.element_to_be_clickable((By.XPATH, button_text))
                            )
                            self.log_with_time(f'✅ Found token button with text: {button_text}')
                            break
                        except TimeoutException:
                            continue
                    
                    if not get_token_button:
                        # Check if there's a cooldown message
                        page_source = self.driver.page_source
                        if 'Chờ' in page_source and 'nữa' in page_source:
                            self.log_with_time('⏰ Key is on cooldown')
                            return {'success': False, 'error': 'Key is on cooldown', 'cooldown': True}
                        
                        # Debug: Log page content
                        self.log_with_time('📄 Debug: Looking for any buttons...')
                        buttons = self.driver.find_elements(By.TAG_NAME, 'button')
                        for i, btn in enumerate(buttons[:5]):  # Check first 5 buttons
                            try:
                                text = btn.text.strip()
                                if text:
                                    self.log_with_time(f'📄 Button {i+1}: "{text}"')
                            except:
                                pass
                        
                        raise TimeoutException("No token button found")
                    
                    self.log_with_time('✅ Found "Lấy Token" button')
                    
                    # Click the button
                    get_token_button.click()
                    self.log_with_time('🔄 Clicking "Lấy Token" button...')
                    time.sleep(5)
                    
                    # Look for token in page content
                    page_source = self.driver.page_source
                    
                    # Look for JWT tokens
                    import re
                    jwt_patterns = [
                        r'eyJ[A-Za-z0-9+/=]{50,}',  # Full JWT
                        r'eyJ[A-Za-z0-9+/=]{100,}', # Very long JWT
                    ]
                    
                    for pattern in jwt_patterns:
                        matches = re.findall(pattern, page_source)
                        if matches:
                            token = matches[0]
                            self.log_with_time(f'🎉 Found token: {token[:30]}...')
                            return {'success': True, 'token': token, 'method': 'selenium'}
                    
                    # If no JWT found, look for any long alphanumeric string
                    long_strings = re.findall(r'[a-zA-Z0-9]{30,}', page_source)
                    if long_strings:
                        # Use the longest string
                        token = max(long_strings, key=len)
                        if len(token) > 20:
                            self.log_with_time(f'🎉 Found potential token: {token[:30]}...')
                            return {'success': True, 'token': token, 'method': 'selenium'}
                    
                    self.log_with_time('⚠️ No token found after clicking button')
                    return {'success': False, 'error': 'No token found after clicking'}
                    
                except TimeoutException:
                    self.log_with_time('❌ "Lấy Token" button not found')
                    return {'success': False, 'error': 'Lấy Token button not found'}
            else:
                self.log_with_time('❌ Login failed, still on login page')
                return {'success': False, 'error': 'Login failed'}
                
        except Exception as e:
            self.log_with_time(f'❌ Token fetching failed: {e}')
            return {'success': False, 'error': str(e)}

    def fetch_token(self):
        """Fetch token using Selenium"""
        if self.is_fetching:
            return {'success': False, 'error': 'Already fetching token'}
        
        # Check cooldown
        if time.time() < self.cooldown_until:
            return {'success': False, 'error': 'Key is on cooldown', 'cooldown': True}
        
        self.is_fetching = True
        
        try:
            result = self.login_and_get_token()
            
            if result['success']:
                self.current_token = result['token']
                self.token_info = result
                self.last_update = time.time()
                self.cooldown_until = 0
                self.log_with_time(f'✅ Token fetched successfully: {result["token"][:30]}...')
                self.send_telegram_message(result['token'])
            else:
                if result.get('cooldown'):
                    self.cooldown_until = time.time() + 300  # 5 minutes cooldown
                    self.log_with_time("⏰ Key on cooldown for 5 minutes")
                self.log_with_time(f"❌ Token fetch failed: {result.get('error', 'Unknown error')}")
            
            return result
                
        except Exception as e:
            self.log_with_time(f'❌ Token fetching failed: {e}')
            return {'success': False, 'error': str(e)}
        finally:
            self.is_fetching = False

    def auto_refresh_worker(self):
        while self.auto_refresh:
            try:
                self.log_with_time("🔄 Auto-refresh: Attempting to fetch token...")
                result = self.fetch_token()
                
                if result['success']:
                    self.log_with_time("✅ Auto-refresh: Token fetched successfully!")
                else:
                    if result.get('cooldown'):
                        cooldown_minutes = 5
                        self.cooldown_until = time.time() + (cooldown_minutes * 60)
                        self.log_with_time(f"⏰ Auto-refresh: Key on cooldown for {cooldown_minutes} minutes")
                    else:
                        self.log_with_time(f"❌ Auto-refresh failed: {result.get('error', 'Unknown')}")
                
                # Wait for next refresh
                self.log_with_time(f"⏳ Auto-refresh: Waiting {self.refresh_interval} seconds...")
                time.sleep(self.refresh_interval)
                
            except Exception as e:
                self.log_with_time(f"❌ Auto-refresh error: {e}")
                time.sleep(60)  # Wait 1 minute on error

    def start_auto_refresh(self, interval=900):
        """Start auto-refresh in background"""
        if self.auto_refresh:
            self.log_with_time("⚠️ Auto-refresh already running")
            return False
        
        self.auto_refresh = True
        self.refresh_interval = interval
        self.refresh_thread = threading.Thread(target=self.auto_refresh_worker, daemon=True)
        self.refresh_thread.start()
        self.log_with_time(f"🚀 Auto-refresh started (interval: {interval}s)")
        return True

    def stop_auto_refresh(self):
        """Stop auto-refresh"""
        if not self.auto_refresh:
            self.log_with_time("⚠️ Auto-refresh not running")
            return False
        
        self.auto_refresh = False
        self.log_with_time("🛑 Auto-refresh stopped")
        return True

    def cleanup(self):
        """Cleanup browser resources"""
        if self.driver:
            try:
                self.driver.quit()
                self.log_with_time('🧹 Browser cleaned up')
            except:
                pass

server = SeleniumTokenServer()

@app.route('/')
def home():
    return jsonify({
        'message': 'Selenium Token Server',
        'status': 'running',
        'version': '1.0.0',
        'endpoints': {
            '/token': 'Get current token',
            '/fetch': 'Fetch new token',
            '/status': 'Get server status',
            '/auto-refresh': 'Manage auto-refresh'
        }
    })

@app.route('/health')
def health_check():
    return jsonify({
        'status': 'healthy',
        'threads': threading.active_count(),
        'timestamp': datetime.now().isoformat()
    })

@app.route('/token')
def get_token():
    if server.current_token:
        return jsonify({
            'success': True,
            'token': server.current_token,
            'last_update': server.last_update,
            'method': server.token_info.get('method', 'unknown') if server.token_info else 'unknown'
        })
    else:
        return jsonify({
            'success': False,
            'error': 'No token available'
        })

@app.route('/fetch', methods=['POST', 'GET'])
def fetch_token_endpoint():
    result = server.fetch_token()
    return jsonify(result)

@app.route('/status')
def get_status():
    return jsonify({
        'success': True,
        'status': 'running',
        'current_token': server.current_token is not None,
        'last_update': server.last_update,
        'is_fetching': server.is_fetching,
        'key_id': server.key_id,
        'auto_refresh': server.auto_refresh,
        'cooldown_until': server.cooldown_until
    })

@app.route('/auto-refresh', methods=['POST'])
def start_auto_refresh_endpoint():
    try:
        data = request.get_json() or {}
        interval = data.get('interval', 900)
        
        if server.start_auto_refresh(interval):
            return jsonify({
                'success': True, 
                'message': f'Auto-refresh started with {interval}s interval',
                'interval': interval
            })
        else:
            return jsonify({'success': False, 'error': 'Auto-refresh already running'}), 400
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@app.route('/auto-refresh', methods=['DELETE'])
def stop_auto_refresh_endpoint():
    try:
        if server.stop_auto_refresh():
            return jsonify({'success': True, 'message': 'Auto-refresh stopped'})
        else:
            return jsonify({'success': False, 'error': 'Auto-refresh not running'}), 400
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@app.route('/auto-refresh/status', methods=['GET'])
def auto_refresh_status():
    return jsonify({
        'success': True,
        'auto_refresh': server.auto_refresh,
        'interval': server.refresh_interval,
        'is_running': server.auto_refresh
    })

if __name__ == '__main__':
    port = int(os.getenv('PORT', 8080))
    server.log_with_time(f'🚀 Starting Selenium Token Server on port {port}')
    server.log_with_time(f'🔑 Using key: {server.key_id}')
    
    server.log_with_time('🧪 Testing token fetch on startup...')
    result = server.fetch_token()
    if result['success']:
        server.log_with_time(f'✅ Startup test successful: {result["token"][:30]}...')
    else:
        server.log_with_time(f'⚠️ Startup test failed: {result.get("error", "Unknown error")}')
    
    auto_refresh_env = os.getenv('AUTO_REFRESH', 'true').lower()
    if auto_refresh_env in ('1', 'true', 'yes', 'on'):
        try:
            interval_env = int(os.getenv('AUTO_REFRESH_INTERVAL', '900'))
        except ValueError:
            interval_env = 900
        started = server.start_auto_refresh(interval=interval_env)
        if started:
            server.log_with_time(f'🔄 Auto-refresh enabled (interval: {interval_env}s)')
        else:
            server.log_with_time('⚠️ Auto-refresh was already running')
    
    try:
        app.run(host='0.0.0.0', port=port, debug=False)
    finally:
        server.cleanup()

File: smart-token-server.js
const express = require('express');
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

const app = express();
const PORT = process.env.PORT || 8080;

// Global variables
let currentToken = null;
let tokenInfo = null;
let lastUpdate = 0;
let consecutiveErrors = 0;
let isFetching = false;

// Configuration
const KEY_ID = process.env.KEY_ID || 'your-key-here';
const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const TELEGRAM_CHAT_ID = process.env.TELEGRAM_CHAT_ID;

// UI Detection patterns
const UI_PATTERNS = {
    inputSelectors: [
        'input[type="text"]',
        'input[type="password"]',
        'input[placeholder*="key"]',
        'input[placeholder*="Key"]',
        'input[name*="key"]',
        'input[name*="Key"]',
        'input[class*="key"]',
        'input[class*="Key"]',
        'input[id*="key"]',
        'input[id*="Key"]',
        'input[aria-label*="key"]',
        'input[aria-label*="Key"]',
        'input'
    ],
    buttonSelectors: [
        'button[type="submit"]',
        'button:contains("Get Token")',
        'button:contains("Lấy Token")',
        'button:contains("Token")',
        'button:contains("Submit")',
        'button:contains("Gửi")',
        'button:contains("Generate")',
        'button:contains("Create")',
        'button.btn-primary',
        'button.btn',
        'button[class*="submit"]',
        'button[class*="generate"]',
        'button[class*="create"]',
        'button'
    ],
    tokenSelectors: [
        '.token-result',
        '.result',
        '[class*="token"]',
        '[class*="result"]',
        '[class*="key"]',
        '[class*="auth"]',
        '[id*="token"]',
        '[id*="key"]',
        '[data-token]',
        '[data-key]',
        'pre',
        'code',
        '.output',
        '.response',
        '.token',
        '.key',
        '.auth',
        '.api-key',
        '.access-token'
    ]
};

// Logging function
function logWithTime(message) {
    const now = new Date();
    const timeStr = now.toLocaleTimeString('vi-VN', { 
        timeZone: 'Asia/Ho_Chi_Minh',
        hour12: false 
    });
    const dateStr = now.toLocaleDateString('vi-VN');
    console.log(`[${timeStr} ${dateStr}] ${message}`);
}

// Browser launch function
async function launchBrowser() {
    const strategies = [
        // Strategy 1: Use system Chromium (fastest)
        {
            headless: true,
            executablePath: '/usr/bin/chromium',
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage',
                '--disable-gpu',
                '--disable-web-security',
                '--disable-features=VizDisplayCompositor',
                '--disable-background-timer-throttling',
                '--disable-backgrounding-occluded-windows',
                '--disable-renderer-backgrounding',
                '--single-process'
            ]
        },
        // Strategy 2: Use system Chrome
        {
            headless: true,
            executablePath: '/usr/bin/chromium-browser',
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage'
            ]
        },
        // Strategy 3: Use Google Chrome
        {
            headless: true,
            executablePath: '/usr/bin/google-chrome',
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage'
            ]
        }
    ];

    for (let i = 0; i < strategies.length; i++) {
        try {
            logWithTime(`🔄 Trying browser launch strategy ${i + 1}...`);
            const browser = await puppeteer.launch(strategies[i]);
            logWithTime(`✅ Browser launched successfully with strategy ${i + 1}`);
            return browser;
        } catch (error) {
            logWithTime(`❌ Strategy ${i + 1} failed: ${error.message}`);
            if (i === strategies.length - 1) throw error;
        }
    }
}

// UI Detection function
async function detectUI(page) {
    logWithTime('🔍 Detecting UI elements...');
    
    const detection = await page.evaluate((patterns) => {
        const result = {
            inputs: [],
            buttons: [],
            tokens: [],
            forms: [],
            allElements: []
        };
        
        // Detect inputs
        patterns.inputSelectors.forEach(selector => {
            try {
                const elements = document.querySelectorAll(selector);
                elements.forEach(el => {
                    result.inputs.push({
                        selector,
                        type: el.type,
                        placeholder: el.placeholder,
                        name: el.name,
                        id: el.id,
                        className: el.className,
                        visible: el.offsetParent !== null
                    });
                });
            } catch (e) {}
        });
        
        // Detect buttons
        patterns.buttonSelectors.forEach(selector => {
            try {
                const elements = document.querySelectorAll(selector);
                elements.forEach(el => {
                    result.buttons.push({
                        selector,
                        text: el.textContent.trim(),
                        type: el.type,
                        className: el.className,
                        id: el.id,
                        visible: el.offsetParent !== null
                    });
                });
            } catch (e) {}
        });
        
        // Detect token elements
        patterns.tokenSelectors.forEach(selector => {
            try {
                const elements = document.querySelectorAll(selector);
                elements.forEach(el => {
                    const text = el.textContent.trim();
                    if (text && text.length > 10) {
                        result.tokens.push({
                            selector,
                            text: text.substring(0, 100),
                            className: el.className,
                            id: el.id,
                            visible: el.offsetParent !== null
                        });
                    }
                });
            } catch (e) {}
        });
        
        // Detect forms
        const forms = document.querySelectorAll('form');
        forms.forEach(form => {
            result.forms.push({
                action: form.action,
                method: form.method,
                className: form.className,
                id: form.id
            });
        });
        
        // Get all interactive elements
        const allInteractive = document.querySelectorAll('input, button, select, textarea, a[href]');
        allInteractive.forEach(el => {
            result.allElements.push({
                tag: el.tagName,
                type: el.type,
                text: el.textContent.trim().substring(0, 50),
                className: el.className,
                id: el.id,
                visible: el.offsetParent !== null
            });
        });
        
        return result;
    }, UI_PATTERNS);
    
    logWithTime(`📊 UI Detection Results:`);
    logWithTime(`  - Inputs: ${detection.inputs.length}`);
    logWithTime(`  - Buttons: ${detection.buttons.length}`);
    logWithTime(`  - Token elements: ${detection.tokens.length}`);
    logWithTime(`  - Forms: ${detection.forms.length}`);
    logWithTime(`  - All interactive elements: ${detection.allElements.length}`);
    
    return detection;
}

// Smart token extraction
async function extractTokenSmart(page, uiDetection) {
    logWithTime('🧠 Smart token extraction...');
    
    // Method 1: Check existing token elements
    for (const tokenEl of uiDetection.tokens) {
        if (tokenEl.visible && tokenEl.text.length > 15) {
            logWithTime(`✅ Found token in element: ${tokenEl.selector}`);
            return tokenEl.text;
        }
    }
    
    // Method 2: Check localStorage/sessionStorage
    const storageToken = await page.evaluate(() => {
        const keys = [...Object.keys(localStorage), ...Object.keys(sessionStorage)];
        for (const key of keys) {
            if (key.toLowerCase().includes('token') || key.toLowerCase().includes('key') || key.toLowerCase().includes('auth')) {
                const value = localStorage.getItem(key) || sessionStorage.getItem(key);
                if (value && value.length > 15) {
                    return value;
                }
            }
        }
        return null;
    });
    
    if (storageToken) {
        logWithTime(`✅ Found token in storage: ${storageToken.substring(0, 20)}...`);
        return storageToken;
    }
    
    // Method 3: Check window variables
    const windowToken = await page.evaluate(() => {
        const tokenVars = ['token', 'apiKey', 'api_key', 'auth', 'authToken', 'accessToken', 'key'];
        for (const varName of tokenVars) {
            if (window[varName] && typeof window[varName] === 'string' && window[varName].length > 15) {
                return window[varName];
            }
        }
        return null;
    });
    
    if (windowToken) {
        logWithTime(`✅ Found token in window variables: ${windowToken.substring(0, 20)}...`);
        return windowToken;
    }
    
    // Method 4: Analyze page text for token patterns
    const pageText = await page.evaluate(() => {
        return document.body.innerText || document.body.textContent || '';
    });
    
    const tokenPatterns = [
        /[A-Za-z0-9]{20,}/g,
        /[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}/g,
        /[A-Za-z0-9+/]{20,}={0,2}/g,
        /[A-Za-z0-9._-]{20,}/g
    ];
    
    for (const pattern of tokenPatterns) {
        const matches = pageText.match(pattern);
        if (matches) {
            for (const match of matches.slice(0, 5)) {
                if (match.length >= 15) {
                    logWithTime(`✅ Found potential token in text: ${match.substring(0, 20)}...`);
                    return match;
                }
            }
        }
    }
    
    return null;
}

// Main token fetching function
async function getTokenFromWebsite() {
    if (isFetching) {
        throw new Error('Already fetching');
    }
    
    isFetching = true;
    let browser = null;
    
    try {
        logWithTime('🔑 Getting token from website...');
        
        // Launch browser
        browser = await launchBrowser();
        const page = await browser.newPage();
        
        // Set viewport and user agent
        await page.setViewport({ width: 1920, height: 1080 });
        await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
        
        // Navigate to website
        await page.goto('https://key-token.com/', {
            waitUntil: 'networkidle0',
            timeout: 60000
        });
        
        logWithTime('⏳ Waiting for page to load...');
        await new Promise(resolve => setTimeout(resolve, 5000));
        
        // Detect UI elements
        const uiDetection = await detectUI(page);
        
        // Try to extract token directly
        const directToken = await extractTokenSmart(page, uiDetection);
        if (directToken) {
            logWithTime(`🎉 Token found directly: ${directToken.substring(0, 20)}...`);
            return { success: true, token: directToken, method: 'direct' };
        }
        
        // If no direct token, try to fill form
        if (uiDetection.inputs.length > 0 && uiDetection.buttons.length > 0) {
            logWithTime('📝 Attempting to fill form...');
            
            // Find best input
            const bestInput = uiDetection.inputs.find(input => 
                input.visible && (
                    input.placeholder?.toLowerCase().includes('key') ||
                    input.name?.toLowerCase().includes('key') ||
                    input.id?.toLowerCase().includes('key')
                )
            ) || uiDetection.inputs.find(input => input.visible);
            
            if (bestInput) {
                await page.type(bestInput.selector, KEY_ID);
                logWithTime(`✅ Filled input: ${bestInput.selector}`);
                
                // Find best button
                const bestButton = uiDetection.buttons.find(button => 
                    button.visible && (
                        button.text.toLowerCase().includes('token') ||
                        button.text.toLowerCase().includes('submit') ||
                        button.text.toLowerCase().includes('generate')
                    )
                ) || uiDetection.buttons.find(button => button.visible);
                
                if (bestButton) {
                    await page.click(bestButton.selector);
                    logWithTime(`✅ Clicked button: ${bestButton.selector}`);
                    
                    // Wait for response
                    await new Promise(resolve => setTimeout(resolve, 5000));
                    
                    // Try to extract token again
                    const newUiDetection = await detectUI(page);
                    const newToken = await extractTokenSmart(page, newUiDetection);
                    
                    if (newToken) {
                        logWithTime(`🎉 Token found after form submission: ${newToken.substring(0, 20)}...`);
                        return { success: true, token: newToken, method: 'form' };
                    }
                }
            }
        }
        
        // Take screenshot for debugging
        try {
            await page.screenshot({ path: '/tmp/debug-screenshot.png' });
            logWithTime('📸 Debug screenshot saved');
        } catch (e) {
            logWithTime(`📸 Screenshot failed: ${e.message}`);
        }
        
        throw new Error('No token found with any method');
        
    } catch (error) {
        logWithTime(`❌ Error getting token: ${error.message}`);
        throw error;
    } finally {
        isFetching = false;
        if (browser) {
            await browser.close();
        }
    }
}

// Auto refresh function
async function autoRefresh() {
    try {
        logWithTime('🔄 Auto token refresh triggered');
        const result = await getTokenFromWebsite();
        
        if (result.success) {
            currentToken = result.token;
            tokenInfo = {
                token: result.token,
                timestamp: new Date().toISOString(),
                source: result.method,
                method: result.method
            };
            lastUpdate = Date.now();
            consecutiveErrors = 0;
            logWithTime(`✅ Token updated successfully via ${result.method}`);
        }
    } catch (error) {
        consecutiveErrors++;
        logWithTime(`❌ Auto refresh failed: ${error.message}`);
    }
}

// Express routes
app.use(express.json());

app.get('/health', (req, res) => {
    res.json({ 
        status: 'healthy', 
        timestamp: new Date().toISOString(),
        uptime: process.uptime()
    });
});

app.get('/token', (req, res) => {
    if (currentToken) {
        res.json({
            success: true,
            token: currentToken,
            info: tokenInfo,
            lastUpdate: new Date(lastUpdate).toISOString()
        });
    } else {
        res.status(404).json({
            success: false,
            message: 'No token available'
        });
    }
});

app.get('/status', (req, res) => {
    res.json({
        status: 'running',
        hasToken: !!currentToken,
        lastUpdate: lastUpdate ? new Date(lastUpdate).toISOString() : null,
        consecutiveErrors,
        isFetching,
        uptime: process.uptime()
    });
});

app.post('/refresh', async (req, res) => {
    try {
        await autoRefresh();
        res.json({ success: true, message: 'Refresh completed' });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

app.post('/auto-refresh', async (req, res) => {
    if (isFetching) {
        return res.status(429).json({ success: false, message: 'Already fetching' });
    }
    
    try {
        await autoRefresh();
        res.json({ success: true, message: 'Auto refresh completed' });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Telegram notification function
async function sendTelegramMessage(message) {
    if (!TELEGRAM_BOT_TOKEN || !TELEGRAM_CHAT_ID) {
        logWithTime('⚠️ Telegram credentials not set');
        return;
    }
    
    try {
        const response = await fetch(`https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/sendMessage`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                chat_id: TELEGRAM_CHAT_ID,
                text: message,
                parse_mode: 'HTML'
            })
        });
        
        if (response.ok) {
            logWithTime('✅ Telegram message sent');
        } else {
            logWithTime(`❌ Telegram send failed: ${response.status}`);
        }
    } catch (error) {
        logWithTime(`❌ Telegram error: ${error.message}`);
    }
}

app.post('/send-token', async (req, res) => {
    if (!currentToken) {
        return res.status(404).json({ success: false, message: 'No token available' });
    }
    
    const message = `🔑 <b>Token Update</b>\n\n` +
                   `Token: <code>${currentToken}</code>\n` +
                   `Method: ${tokenInfo?.method || 'unknown'}\n` +
                   `Time: ${new Date().toISOString()}`;
    
    await sendTelegramMessage(message);
    res.json({ success: true, message: 'Token sent to Telegram' });
});

// Start server
app.listen(PORT, () => {
    logWithTime(`🚀 Smart Token Server started on port ${PORT}`);
    logWithTime('📡 Endpoints:');
    logWithTime('   GET  /health - Health check');
    logWithTime('   GET  /token - Get current token');
    logWithTime('   GET  /status - Get server status');
    logWithTime('   POST /refresh - Force refresh token');
    logWithTime('   POST /auto-refresh - Auto refresh (for cron)');
    logWithTime('   POST /send-token - Send token to Telegram');
    
    // Initial token fetch
    setTimeout(() => {
        logWithTime('🔄 Initial token fetch...');
        autoRefresh();
    }, 2000);
    
    // Auto refresh every 5 minutes
    setInterval(autoRefresh, 5 * 60 * 1000);
});

// Graceful shutdown
process.on('SIGTERM', () => {
    logWithTime('🛑 Shutting down gracefully...');
    process.exit(0);
});

process.on('SIGINT', () => {
    logWithTime('🛑 Shutting down gracefully...');
    process.exit(0);
});
